### ISSUE TITLE
Enum. Allowed characters

### ISSUE THREAD

**Issue by MCON** [Folkloreatelier/laravel-graphql#153]
One of our programmers tried to use enum. It refused to work if it's keys had dashes. I think enum should work with any symbols no?

**Comment by vladar** [comment]
This is done for compatibility with [graphql-js](https://github.com/graphql/graphql-js/blob/master/src/type/definition.js#L989), they use regex `/^[_a-zA-Z][_a-zA-Z0-9]*$/` for enum values.

If you think it shouldn't be the case, raise an issue on graphql-js repo.

**Comment by MCON** [comment]
Created issue in their project:
https://github.com/graphql/graphql-js/issues/936

I think it is crucial to allow all characters.
We can make input filtering, nice schemas etc. then.
What you think?

**Comment by vladar** [comment]
We will keep compatibility with reference implementation anyway (graphql-js). 

But just curious. Can you share your usage scenario? What are you trying to achieve that you need dash and other chars vs just using alnum+underscore?

**Comment by MCON** [comment]
We have multiple forms. They have external ids for integration. We want this external ids to be listed in enum. Those who integrate with our system would see allowed values in graphiql.


```

    // arguments to filter query
    public function args()
    {
        $formEnum = new EnumType(['name' => 'form', 'values' => FormScript::getAllFormValues()]);

        return [
            'id'    => [
                'name' => 'id',
                'type' => Type::int()
            ],
            'name'  => [
                'name' => 'name',
                'type' => Type::string()
            ],
            'form'  => [
                'name' => 'form',
                'type' => $formEnum
            ],
```

**Comment by vladar** [comment]
Usually something stored in database in not the best candidate for enum. You should probably consider using other types for that.

But if you do want to expose it as enum, `graphql-php` library (which is underlay library for `laravel-graphql`) allows you to convert between your own internal values and names exposed via graphql introspection.

Consider following code ([example from docs](http://webonyx.github.io/graphql-php/type-system/enum-types/#field-resolution)):

```php
use GraphQL\Type\Definition\EnumType;
use GraphQL\Type\Definition\ObjectType;

$episodeEnum = new EnumType([
    'name' => 'Episode',
    'description' => 'One of the films in the Star Wars Trilogy',
    'values' => [
        'NEWHOPE' => [
            'value' => 4,
            'description' => 'Released in 1977.'
        ],
        'EMPIRE' => [
            'value' => 5,
            'description' => 'Released in 1980.'
        ],
        'JEDI' => [
            'value' => 6,
            'description' => 'Released in 1983.'
        ],
    ]
]);

$heroType = new ObjectType([
    'name' => 'Hero',
    'fields' => [
        'appearsIn' => [
            'type' => $episodeEnum,
            'resolve' => function() {
                return 5; // Actual entry in response will be 'appearsIn' => 'EMPIRE'
            }
        ]
    ]
])
```

As you see each enum entry may have `value`. It can be anything you want. This is your internal vlaue. `name` on the other hand (in this example defined as array key) is what is exposed to graphql clients via introspection.

So in your resolvers you can return internal values and they will be converted to enum names in response. Reverse is true when you accept enum as input, graphql will expect the name, but will convert it to your internal value when passing to resolver arguments.

I am not sure though how to express it with `laravel-graphql`.

**Comment by MCON** [comment]
No-no you propose me to hardcode something to make it verbose. This is absolutely not something i want.
